Reset Modules as an Xcode local package with tests in CI

mokagio opened the pull request:
## Description

https://github.com/wordpress-mobile/WordPress-iOS/pull/24511 successfully addressed the issue of duplicated packages in the target selection.

Unfortunately, it resulted in all local packages (those from `Modules/` that is) no longer being accessible to the unit tests runner—see internal discussion in p1746169066574499/1745343713.430699-slack-C08HQR4C2TS

This PR propose a different fix, achieved by re-adding `Modules/` **as a local _package_ not a package _dependency_**, following the instructions from https://developer.apple.com/documentation/xcode/organizing-your-code-with-local-packages.

This setup maintains the single result in the target search...

<img width="495" alt="image" src="https://github.com/user-attachments/assets/ad8d1b7a-cd2d-4204-b9b8-40adeeb0e512" />

...while also allowing for `Modules/` to run as dedicated schemes or as part of a test plan

<img width="1051" alt="image" src="https://github.com/user-attachments/assets/fb682cbc-b4e6-43d8-b154-ce100f1df4e3" />

<img width="710" alt="image" src="https://github.com/user-attachments/assets/62b449ce-46a1-4058-a091-90dc04736529" />

Also notice, we have so many tests (and warnings when running them!) that the logs don't show the first tests that run, but if you check the XML report, you'll see that indeed we are running all previously disabled tests

![image](https://github.com/user-attachments/assets/bb7bd2a2-c564-431d-a0cd-077fcaeae70e)


## Next steps

- Rename `WordPressUnitTests.xctestplan` to `Keystone.xctestplan`—to do in dedicated PR to avoid diff noise here
- Ensure all module targets that have tests run as part as `Keystone.xctestplan`
- Add dedicated schemes for focused package development

---

https://linear.app/a8c/issue/AINFRA-428/investigate-wjr-setup-that-addresses-duplicated-modules-while

mokagio committed:
Add a dummy local package—We'll try to put the Modules there next

mokagio committed:
Replace ModulesAlterantive with Modules

mokagio committed:
Re-enable all the test targets in WordPressUnitTests

dangermattic commented:
<!--
  0 Errors
  1 Warning: <code>Modules/Package.swift</c...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages as appropriate to your project setup (e.g. in Xcode or by running <code>swift package resolve</code>).</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mokagio committed:
Delete unnecessary OCLint scheme

Given OCLint is something we should drop soon, because it's only for
Objective-C and we want to reduce that code surface.

mokagio committed:
Configure WordPressFlux scheme to run tests

mokagio commented on Modules/Package.swift:
```diff
@@ -169,7 +170,12 @@ let package = Package(
         ),
         .testTarget(name: "JetpackStatsWidgetsCoreTests", dependencies: [.target(name: "JetpackStatsWidgetsCore")], swiftSettings: [.swiftLanguageMode(.v5)]),
         .testTarget(name: "DesignSystemTests", dependencies: [.target(name: "DesignSystem")], swiftSettings: [.swiftLanguageMode(.v5)]),
-        .testTarget(name: "WordPressFluxTests", dependencies: ["WordPressFlux"], swiftSettings: [.swiftLanguageMode(.v5)]),
+        .testTarget(
+            name: "WordPressFluxTests",
+            dependencies: ["WordPressFlux"],
+            exclude: ["WordPressFluxTests.xctestplan"],
+            swiftSettings: [.swiftLanguageMode(.v5)]
+        ),
```
I added this to see how `WordPressFluxTests` would run tests with standalone scheme.

Interestingly, I tried using a copy-paste friendly `exclude: ["*.xctestplan"],` pattern, but it failed to find the file.

mokagio replied:
My autogenerated scheme picks up the test plan:

<img width="956" alt="image" src="https://github.com/user-attachments/assets/f998752d-c899-42fc-81e0-c08c123fb8d1" />

However, I got an unlock prompt when I edited it (but no `xcsheme` change in the project, possibly because the scheme might be in `Modules/.swiftpm`) so I'd be curious to know if you see it too...

mokagio commented on WordPress/WordPress.xcodeproj/project.pbxproj:
```diff
@@ -7085,13 +7084,6 @@
-			relativePath = ../Modules;
-		};
-/* End XCLocalSwiftPackageReference section */
-
```
I think the difference between this and the previous setup (pre https://github.com/wordpress-mobile/WordPress-iOS/pull/24511) is that we no longer have _any_ local swift package reference as a dependency for WordPress. 

My guess is that, whichever way it ended up like that, the duplicated reference—as a local module in the workspace and as a package reference—is what resulted in the duplicated entries shown in the #24511 description.

kean commented on Modules/Package.swift:
```diff
@@ -6,6 +6,7 @@ let package = Package(
     name: "Modules",
     platforms: [
         .iOS(.v16),
+        .macOS(.v14), // Here just to avoid errors when building directly with `swift build`
```
Is this addition necessary? Most of our packages do not support macOS and won't compile with it as a target.

mokagio replied:
I find it handy to have just so `swift build` throws legitimate errors with the code rather than with the versions. 

In particular, it was necessary to run the WordPressFlux tests via `swift test`. However, as I try to reproduce, I can't get past some UIKit errors, which I find odd when focused on WordPressFlux, but don't have bandwidth to investigate.

Removed it.

kean replied:
Thanks! Yeah, I'd remove it since it doesn't support macOS.

kean commented on WordPress.xcworkspace/contents.xcworkspacedata:
```diff
@@ -1,6 +1,9 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <Workspace
    version = "1.0">
+   <FileRef
```
Let's add `Modules` to a project and not a workspace. We we going to remove the workspace.

mokagio replied:
Done.

mokagio replied:
> We we going to remove the workspace.

I started looking at it https://github.com/wordpress-mobile/WordPress-iOS/pull/24534

kean reviewed (commented):
This looks good. Let's switch back to the standard "drag-n-drop" approach. I left a couple of small comments.

mokagio committed:
Move Modules package reference from workspace to project

mokagio committed:
Use the shared setup CI script in the complete code freeze step

mokagio committed:
Add workaround for package path support missing

kean commented on .gitignore:
```diff
@@ -125,12 +125,6 @@ WordPress/InfoPlist-internal.h
-#
-# Ignore the Package.resolved file. The packages source of truth is the
-# resolved file generated by the Xcode workspace.
-Modules/Package.resolved
```
I thought you had to use `...ace/xcshareddata/swiftpm/Package.resolved` managed by Xcode? `Modules/Package.resolved` isn't, or is it? How did you get it to work? When you open the Xcode project, does it use pins from `Modules/Package.resolved`?

```
...ace/xcshareddata/swiftpm/Package.resolved → Modules/Package.resolved
```

mokagio replied:
Apologies @kean this change surprised me a bit and I should have preempted your question writing about it.

## Before

`Modules` was part of the workspace. When Xcode detected changes in `Modules/Package.swift`, it triggered the packages resolution and updated the resolved file in the workspace `xcshareddata` folder.

## After

`Modules` is part of the project, but is not a dependency of the project:

![image](https://github.com/user-attachments/assets/9abdaa78-f7cd-4b6f-97d3-e0b98cc9acbd)

When Xcode detects a change in `Modules/Package.swift`, it triggers the packages resolution and updates `Modules/Package.resolved`.

I guess that if we were to add additional package dependencies _to the project_ Xcode would then re-introduce the  workspace resolved file.

mokagio replied:
> When Xcode detects a change in Modules/Package.swift, it triggers the packages resolution and updates Modules/Package.resolved.

I verified this by switching `color-studio` from pointing to a branch to a commit. The screenshot below shows Xcode highlighting the change in the gutter and the package resolution running.

![image](https://github.com/user-attachments/assets/01846ab0-6031-4e30-8238-915d0775ce38)

35e360bb81 tracks the changes and the `.resolved` update.

mokagio committed:
Point `color-studio` to a commit instead of a branch

Locking dependencies to commits makes updating them a bit more
burdensome, but at least it makes the update intentional and not
automatic and surprising.

This change was done to test Xcode's behavior in monitoring
`Modules/Package.swift` and updating `Modules/Package.resolved`.

See
https://github.com/wordpress-mobile/WordPress-iOS/pull/24518#discussion_r2079810655
